src/gt4py/analysis/passes.py

```diff
@@ -1290,6 +1290,89 @@ def apply(cls, transform_data: TransformData) -> None:
+                temp_field: None for temp_field in node.temporary_fields
+            }
+            self.visit(node)
+            return set(self.reduced_fields.keys())
+
+        def visit_MultiStage(self, node: gt_ir.MultiStage) -> None:
+            self.iteration_order = node.iteration_order
+            self.generic_visit(node)
+
+        def visit_ApplyBlock(self, node: gt_ir.ApplyBlock) -> None:
+            self.interval = node.interval
+            self.generic_visit(node)
+
+        def visit_FieldRef(self, node: gt_ir.FieldRef) -> None:
+            field_name = node.name
+            if field_name in self.reduced_fields:
+                if self.iteration_order == gt_ir.IterationOrder.PARALLEL:
+                    # Do not reduce fields accessed from parallel k-intervals
+                    self.reduced_fields.pop(field_name)
+                else:
+                    offsets: List[int] = list(node.offset.values())
+                    if offsets[-1] != 0:
+                        # Do not reduce fields with k-offsets
+                        self.reduced_fields.pop(field_name)
+                    else:
+                        # Do not reduce fields that are accessed across different k-intervals
+                        interval = self.reduced_fields[field_name]
+                        if interval is not None and interval != self.interval:
+                            self.reduced_fields.pop(field_name)
+                        else:
+                            self.reduced_fields[field_name] = self.interval
+
+    class StorageReducer(gt_ir.IRNodeMapper):
+        @classmethod
+        def apply(cls, node: gt_ir.StencilImplementation, reduced_fields: Set[str]) -> None:
+            instance = cls(reduced_fields)
+            return instance(node)
+
+        def __init__(self, reduced_fields: Set[str]):
+            self.reduced_fields = reduced_fields
+
+        def __call__(self, node: gt_ir.StencilImplementation) -> gt_ir.StencilImplementation:
+            assert isinstance(node, gt_ir.StencilImplementation)
+            return self.visit(node)
+
+        def visit_StencilImplementation(
+            self, path: tuple, node_name: str, node: gt_ir.StencilImplementation
+        ) -> gt_ir.StencilImplementation:
+            self.iir = node
+            for field_name in self.reduced_fields:
+                assert field_name in node.temporary_fields, "Tried to reduce API field to 2D."
+                field_decl = node.fields[field_name]
+                field_decl.axes.pop()
+            return self.generic_visit(path, node_name, node)
+
+        def visit_FieldRef(
+            self, path: tuple, node_name: str, node: gt_ir.FieldRef
+        ) -> Tuple[bool, Union[gt_ir.FieldRef, gt_ir.VarRef]]:
+            if node.name in self.reduced_fields:
+                field_decl = self.iir.fields[node.name]
+                node.offset = {axis: node.offset[axis] for axis in field_decl.axes}
+            return True, node
```

Comment:
I think it's safer to return a new node instead of modifying the existing one

Reply:
Also implemented.